File: C++/Leetcode/Linked List/LC19-RemoveNthNodeFromEndOfList.cpp
// LC19: Remove Nth Node From End of List

// Definition for singly-linked list.
struct ListNode {
	int val;
	ListNode *next;
	ListNode() : val(0), next(nullptr) {}
	ListNode(int x) : val(x), next(nullptr) {}
	ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // Use 1 ptr to get (n-1) offset from head and another to get (n-1)st position
        // by advancing both pointers until the 1st ptr reaches null
        ListNode* offset = head;
        ListNode* slow = head;

        for (int i = 0; i < n; i++) {
            offset = offset->next;
        }

        if (!offset) return head->next; // If n == head.size, just return tail of list

        while (offset->next) {
            offset = offset->next;
            slow = slow->next;
        }

        // Perform deletion (constant time)
        slow->next = slow->next->next;

        return head;
    }
};

File: go/Leetcode/Linked List/LC19-RemoveNthNodeFromEndOfList.go
package leetcode

// Definition for singly-linked list.
type ListNode struct {
    Val int
    Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
    // use slow and fast ptr method to find (n-1)st node, where fast is n nodes offset 
    offset := head
    slow := head

    for n > 0 {
        offset = offset.Next
        n--
    }

    // catch edge case where n is the size of the whole list
    if offset == nil {
        return head.Next
    }

    // shift both until slow is at (n-1)st node
    for offset.Next != nil {
        offset = offset.Next
        slow = slow.Next
    }

    // perform deletion
    slow.Next = slow.Next.Next

    return head
}


